src/App.tsx

```diff
@@ -24,7 +26,12 @@ function App() {
           <img src={reactLogo} className="logo react" alt="React logo" />
         </a>
       </div>
-      <h1>Vite + React + Vanilla-Extract</h1>
+      <SprinklesBox p={"2xl"} m={["sm", "lg", "xl", "2xl"]}>
```

Comment:
[info] @riccardosilvi Non mi è chiaro devo vengono definiti i valori di questi breakpoint.

Reply:
in src/styles/sprinkles.css.ts viene utilizzata l'api `defineProperties`.
```ts
// con un'oggetto statico definisco tutte le proprietà per cui poi tramite defineProperties definirò come "responsive"
// esempio 
// per display per ogni breakpoint potrò scegliere tra `none`, `block`, `inline`, `inline-block`, `flex` ( e non altri potenziali valori che sarebbero nativamente supportati dalla css property display)
export const responsiveSupportedProps = {
  position: [`relative`],
  display: [`none`, `block`, `inline`, `inline-block`, `flex`],
  alignItems: [`flex-start`, `center`, `flex-end`, `baseline`],
  justifyContent: [`flex-start`, `center`, `flex-end`, `space-between`],
  flexDirection: [`row`, `row-reverse`, `column`, `column-reverse`],
  flexWrap: [`wrap`, `nowrap`],
  padding: spacing,
  paddingTop: spacing,
  paddingBottom: spacing,
  paddingLeft: spacing,
  paddingRight: spacing,
  margin: spacing,
  marginTop: spacing,
  marginBottom: spacing,
  marginLeft: spacing,
  marginRight: spacing,
  columnGap: spacing,
  rowGap: spacing,
} as const;

// importante l'ordine qui, passandolo a defineProperties all'omonima proprietà definisce 
// l'ordine in cui andranno immessi i valori per i breakpoint così specificati
export const responsiveArray = ["xs", "sm", "md", "lg", "xl", "2xl"] as const;

// ecco define properties che accetta delle conditions che nel caso di voler rendere delle regole responsive
// sono i breakpoints
const responsiveProperties = defineProperties({
  conditions:
    transformBreakpoints<Record<keyof typeof BREAKPOINTS, {}>>(BREAKPOINTS),
  //  ecco qui il responsiveArray
  responsiveArray: responsiveArray,
  // xs è il breakpoint di default da cui si applicano le regole
  defaultCondition: "xs" as const,
  // elenco delle proprietà per cui dare la possibilità di passare props con
  // display={{ xs : "block", md: "none" }}
    // display={["block", null, "none"]}
  properties: responsiveSupportedProps,
  // shorthands sono le regole composte
  // per esempio "px" coinvolge padding-left e padding-right
  shorthands: {
    p: [`paddingTop`, `paddingBottom`, `paddingLeft`, `paddingRight`],
    px: [`paddingLeft`, `paddingRight`],
    py: [`paddingTop`, `paddingBottom`],
    m: [`marginTop`, `marginBottom`, `marginLeft`, `marginRight`],
    mx: [`marginLeft`, `marginRight`],
    my: [`marginTop`, `marginBottom`],
  },
});

// Si passa infine tutto a createSprinkles e il risultato è ciò che useremo
// per stilare secondo le regole specificate
export const sprinkles = createSprinkles(responsiveProperties);

// tramite questo export abbiamo modo di estrapolare ed utilizzare type 
export type Sprinkles = Parameters<typeof sprinkles>[0];
```

Poi restando sull'esempio del Box
```tsx

// utilizziamo il type di sprinkles per definire come oltre agli attributi nativi il Box
// accetterà anche tutte le props responsive specificate
export type SprinklesBoxProps = PropsWithChildren<
  React.HTMLAttributes<HTMLDivElement> & Sprinkles
>;

export const SprinklesBox = ({
  className = "",
  children,
  ...rest
}: SprinklesBoxProps) => {
  // sprinkles alla fine genera una classe 
  return (
    <div className={`sprinkles-box ${className} ${sprinkles(rest)}`} {...rest}>
      {children}
    </div>
  );
};

export default SprinklesBox;
```

Per qualsiasi altra info non esitare!

Reply:
@riccardosilvi  Ok. Però continuo a non capire dove vengono settati i valori per le properties definite.

Per esempio come posso vedere quanto viene messo di margin nei casi `sm` e `md`?

Reply:
La sintassi è analoga a quella di styled system ma purtroppo credo il naming possa aver tratto in inganno

```tsx
<SprinklesBox 
  m={["sm", "lg", "xl", "2xl"]}>
```
Analizziamo la prop "m" (omessa "p" a scopo di esempio ma funziona allo stesso modo)

come da `defineProperties` la property "m" rappresenta le quattro regole di margin nella sezione shorthands 
```ts
shorthands: {
 m: [`marginTop`, `marginBottom`, `marginLeft`, `marginRight`],
}
```
Risaliamo quindi a che valori supportano queste regole in `responsiveSupportedProps`
```ts
// le regole di spacing sono estratte da vars che racchiude i valori "themed"
const { spacing } = vars;

export const responsiveSupportedProps = {
 marginTop: spacing,
 marginBottom: spacing,
 marginLeft: spacing,
  marginRight: spacing,
  ...rest
}
```

Andando in src/styles/theme/baseTheme.css.ts troviamo l'oggetto `spacing`

```ts
export const spacing = {
  "6xs": `2px`,
  "5xs": `4px`,
  "4xs": `8px`,
  "3xs": `12px`,
  "2xs": `16px`,
  xs: `24px`,
  sm: `32px`,
  md: `40px`,
  lg: `48px`,
  xl: `56px`,
  "2xl": `64px`,
} as const;

// poi utilizzato all'interno del tema di base che viene poi "esteso" dai temi dark e light
export const baseTheme = {
  spacing,
  ...restRules
};
```

 tenendo a mente che l'array di breakpoints da controfillare coi valori è
```ts
export const responsiveArray = ["xs", "sm", "md", "lg", "xl", "2xl"] as const;
```
ritorniamo al Box,
```tsx
<SprinklesBox 
  m={["sm", "lg", "xl", "2xl"]}>
```
possiamo esprimere il margine in due modi: 
```tsx
// oggetto {[breakpoinName]: propertyValue}
<SprinklesBox 
  m={{
     xs: "sm", // vars.spacing.sm
     sm: "lg",  // vars.spacing.lg
     md: "xl",  // vars.spacing.xl
     lg: "2xl",  // vars.spacing["2xl"]
  }}>
```
tradotto in array i valori che passiamo sono le chiavi di spacing a ciascun breakpoint
```tsx
<SprinklesBox 
  m={["sm", "lg", "xl", "2xl"]}>
```

Avessimo avuto i breakpoint "sm" e "md" uguali ad "xs" e solo da "lg" un change avremmo messo dei null, proprio come facciamo ora con styled system
```tsx
<SprinklesBox 
  m={["sm", null,null, "2xl"]}>
```